### PR TITLE
Change filter on /rails/info/routes to use an actual path regexp from rails

### LINKED ISSUE #18402: rake routes fails when route constraint regex contains comments
When the constraint regex on a route contains comments, `rake routes` fails with the following error:

```
RegexpError: end pattern with unmatched parenthesis: /^\/entry\/((?x-mi:[a-f\d]{8}-[a-f\d]{4}-[1-5]#Version:http:\/\/tools.ietf.org\/html\/rfc4122#section-4.1.3[a-f\d]{3}-[89ab]#Variant:http:\/\/tools.ietf.org\/html\/rfc4122#section-4.1.1[a-f\d]{3}-[a-f\d]{12}))$/
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:52:in `initialize'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:52:in `new'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:52:in `json_regexp'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:133:in `block in collect_routes'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:126:in `collect'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:126:in `collect_routes'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:93:in `format'
	from (irb):4
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/railties-4.1.5/lib/rails/commands/console.rb:90:in `start'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/railties-4.1.5/lib/rails/commands/console.rb:9:in `start'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/railties-4.1.5/lib/rails/commands/commands_tasks.rb:69:in `console'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/railties-4.1.5/lib/rails/commands/commands_tasks.rb:40:in `run_command!'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/railties-4.1.5/lib/rails/commands.rb:17:in `<top (required)>'
	from bin/rails:4:in `require'
	from bin/rails:4:in `<main>'
```

The code in question is [here](https://github.com/alphagov/content-register/blob/2ced4612654a0a18d97b881a6ee52684f76a0c2f/config/routes.rb#L4) and as follows:

```ruby
Rails.application.routes.draw do
  with_options :format => false do |r|
    r.get '/entries' => 'entries#index'
    r.put '/entry/:content_id' => 'entries#update', constraints: { content_id: Entry::UUID_REGEX }

    r.get '/healthcheck' => proc {|env| [200, {}, ["OK"]]}
  end
end
```

The regex is [here](https://github.com/alphagov/content-register/blob/2ced4612654a0a18d97b881a6ee52684f76a0c2f/app/models/entry.rb#L4) and as follows:

```ruby
UUID_REGEX = %r{
    [a-f\d]{8}
    -
    [a-f\d]{4}
    -
    [1-5]   # Version: http://tools.ietf.org/html/rfc4122#section-4.1.3
    [a-f\d]{3}
    -
    [89ab]  # Variant: http://tools.ietf.org/html/rfc4122#section-4.1.1
    [a-f\d]{3}
    -
    [a-f\d]{12}
  }x
```

The line that it fails on in the source is [here](https://github.com/rails/rails/blob/v4.1.5/actionpack/lib/action_dispatch/routing/inspector.rb#L52). I know this isn't the latest version of Rails but that method looks like it hasn't changed at all compared to [master](https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/routing/inspector.rb#L45).

### PR DESCRIPTION
**Pull request by brainopia**
Change filter on `/rails/info/routes` to use an actual path regexp from rails
and not approximate javascript version. Oniguruma supports much more
extensive list of features than javascript regexp engine.

Fixes #18402 (by removing json_regexp which breaks down on complex regexp patterns) 

cc @schneems @rafaelfranca

### CODE REVIEW

**Comment by brainopia** [comment]
Example of filtering on a complex pattern (which breaks `/rails/info/routes` page and `rake routes` on master)

**Comment by arthurnn** [comment]
@schneems what do you think?

**Comment by rohit** [comment]
I've only given a cursory look at the PR but in general I agree with the direction. The conversion of Ruby compatible regexp to JS compatible is fragile and will be prone to errors. Also `rake routes` is an important command that ought not to break just because there is a complex regexp.

Another alternative will be to catch the error during conversion and revert back to original regexp. Basically if you get till the conversion the original regexp is fine but converting it is not working out so print the regex as-is. Of course this leads to inconsistency and might be surprising.

**Comment by schneems** [comment]
Thanks for the work, and sorry for the delay. This is a pretty big change for a pretty small bug. As you noted, this page doesn't have any tests, so i've hesitant to merge in large changes. Moving the processing to the server side is a bit heavy handed. I'm :-1:  on it as is. Can we modify the regex sanitizer to remove comments?

**Comment by brainopia** [comment]
The ruby regexp engine supports much more features than javascript version.
Why try to achieve a working approximation when exact match can be achieved?

If you worry about complex logic, then you shouldn't. I've removed more lines of code than I've added (including a big changelog entry) and the logic is straightforward.

**Comment by rafaelfranca** [comment]
Also we want to make it possible to use the fuzzy match in the rake tasks
so it is better to have to logic at Ruby side.

**Comment by schneems** [comment]
I'm convinced. I'll test out the implementation locally in the next day or so.

Is there anyway to test any of this? Are we doing integration style tests on any of the other internal controllers?

**Comment by rafaelfranca** [comment]
Maybe unit test in the fuzzy match class? But I think we have tests to the
info controller

**Comment by schneems** [comment]
I verified and it works as promised. We should add a test to https://github.com/rails/rails/blob/master/railties/test/rails_info_controller_test.rb to test the results of querying for a partial and full match return as expected.

This commit did break the routing error page: https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb

Basically whenever you get a routing exception you get a very similar view to `rails/info/routes` that lets you debug the problem on the same page. It needs to be updated to use the new server architecture

**Comment by brainopia** [comment]
@schneems Thank you for feedback!

I've fixed the issue with `routing_error.html.erb` page (I was using a relative path in javascript, changed it to absolute to work from arbitrary locations).

And I've added tests.

**Comment by schneems** [comment]
The error page works, your page works. I agree this is cleaner and simpler than my first version. Thanks for the PR, merging in :heart: :heart: :heart:

**Comment by brainopia** [comment]
Thank you :bow:
